### PR TITLE
Crop MTP drafts at the first EOS

### LINKED ISSUE #47912: Assisted/speculative decoding generates past an EOS that is accepted mid-block
# Assisted/speculative decoding generates past an EOS that is accepted mid-block

### System info

- `transformers` 5.15.0 (reproduced on the `v5.15.0` tag)
- `torch` 2.13.0 — the minimal repro is CPU-only
- Originally observed on `meta-models/Muse-Glimmer-30B` + `-assistant` (DFlash), A100 80GB, bf16

### Who can help?

@Cyrilvallez (generate). The DFlash generator came in with #47867.

## Description

`_assisted_decoding` commits `n_matches + 1` tokens per round. When an accepted
token in the middle of that block is an EOS, generation does not stop: the whole
block is appended (and streamed) first, and the stopping check afterwards only
looks at the last committed token. The result is output that continues past EOS
and does not match plain greedy.

It is position-dependent, which makes it look intermittent — an EOS that happens
to land on the target's bonus token stops correctly.

## Minimal reproduction (CPU, tiny random model, no drafter needed)

The bug needs the EOS to be *accepted*, i.e. the target's own argmax must be EOS
at that position. Instead of hunting for such a prompt, run greedy first and then
declare one of the tokens the model actually produced to be the EOS.

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers.generation.candidate_generator import CandidateGenerator
import transformers.generation.utils as gu

NAME, BLOCK, EOS_OFFSET = "hf-internal-testing/tiny-random-LlamaForCausalLM", 8, 3
model = AutoModelForCausalLM.from_pretrained(NAME).eval()
tok = AutoTokenizer.from_pretrained(NAME)
ids = tok("Hello world", return_tensors="pt").input_ids
plen = ids.shape[1]

cont = model.generate(ids, max_new_tokens=BLOCK * 2, do_sample=False,
                      eos_token_id=-1, pad_token_id=0)[0, plen:].tolist()
EOS = cont[EOS_OFFSET]                      # a token the model really emits

class BlockStub(CandidateGenerator):        # proposes the model's own continuation
    requires_model_outputs = False
    def get_candidates(self, input_ids, **kw):
        n = input_ids.shape[1] - plen
        nxt = cont[n:n + BLOCK]
        if not nxt:
            return input_ids, None
        return torch.cat([input_ids, torch.tensor([nxt])], dim=-1), None
    def update_candidate_strategy(self, input_ids, scores, num_matches):
        pass

base = model.generate(ids, max_new_tokens=BLOCK * 2, do_sample=False,
                      eos_token_id=EOS, pad_token_id=0)
gu.GenerationMixin._get_candidate_generator = lambda self, **kw: BlockStub()
spec = model.generate(ids, max_new_tokens=BLOCK * 2, do_sample=False,
                      eos_token_id=EOS, pad_token_id=0, prompt_lookup_num_tokens=BLOCK)

print(base[0, plen:].tolist())   # [30016, 7356, 9023, 25068]                   <- stops at EOS
print(spec[0, plen:].tolist())   # [30016, 7356, 9023, 25068, 22665, ... ]      <- 12 tokens past EOS
```

Observed: plain greedy returns 4 tokens, assisted returns 16 — **12 tokens after
the EOS**.

### On a real model

`meta-models/Muse-Glimmer-30B` + its DFlash drafter, greedy, `max_new_tokens=128`,
prompt `"List the first eight prime numbers, one per line, with no commentary."`:

| | length | EOS index | ends with EOS |
|---|---|---|---|
| plain greedy | 122 | 121 | yes |
| `speculation_type="dflash"` | **128** | 121 | **no** |

The first 122 tokens are identical; the speculative stream then emits
`<|start|>assistant to=self<|message|>We`. Instrumenting the candidate generator
to record `(proposed, num_matches)` per round and locating the EOS in its commit
round explains the intermittency:

- EOS at offset 6 of an 8-token commit (an **accepted proposal**) → ran 6 tokens past it.
- On another prompt, EOS at offset 1 of a 2-token commit (the **bonus** token) → stopped correctly.

## Expected behavior

Greedy assisted decoding should return the same token stream as plain greedy, and
stop at the first EOS. Instead it continues past an EOS that was accepted inside a
committed block, so the two disagree on both content and length.

## Cause

Two things compound, both in `_assisted_decoding`:

1. The committed block is appended and streamed before any stopping check:
   ```python
   input_ids = torch.cat((input_ids, valid_tokens), dim=-1)   # utils.py:3808
   if streamer is not None:
       streamer.put(valid_tokens.cpu())
   ```
2. The check that follows uses the default `new_token_length=1`:
   ```python
   unfinished_sequences = unfinished_sequences & ~stopping_criteria(input_ids, scores)  # utils.py:3859
   ```
   and `EosTokenCriteria` only inspects the last `new_token_length` tokens
   (`stopping_criteria.py:580`):
   ```python
   is_done = torch.isin(input_ids[:, -new_token_length:], self.eos_token_id).any(dim=-1)
   ```

The existing guard at `utils.py:3777` does not cover it: it fires only when the
*entire* candidate block was accepted (`n_matches == candidate_length`), and its
`is_done_candidate` (`utils.py:3696`) is itself computed at `new_token_length=1`.

Classic autoregressive drafters largely avoid this because the draft stops at EOS,
leaving it last. Block proposers (DFlash `block_size` 16, MTP) hit it routinely.

This failure mode is already known in-repo — `PromptLookupCandidateGenerator`
truncates its candidates at the first EOS for exactly this reason
(`candidate_generator.py:1129`):

```python
# remove remaining candidate ids if an "eos" token is found, otherwise the target model may
# accept eos and the rest as valid, thus not stopping generation after "eos"
```

`DFlashTokenCandidateGenerator` has no equivalent.

## Fix

Trim the committed block at the first EOS, mirroring what
`PromptLookupCandidateGenerator` already does to its drafts. That also leaves EOS
as the last token, so the existing criteria fire unchanged:

```python
eos_tensor = generation_config._eos_token_tensor
if eos_tensor is not None and valid_tokens.shape[1] > 1:
    eos_hits = torch.nonzero(torch.isin(valid_tokens[0], eos_tensor.to(valid_tokens.device)))
    if eos_hits.numel() > 0:
        valid_tokens = valid_tokens[:, : eos_hits[0].item() + 1]
        n_matches = valid_tokens.shape[1] - 1
```

placed just after the existing length-budget clamp, recomputing `n_matches` the way
that clamp already does so `number_of_tokens_to_crop` stays consistent.

Note the obvious one-liner — passing `new_token_length=n_matches + 1` to the
stopping criteria at `utils.py:3859`, as `generation_diffusion_gemma.py:1088` does
for its canvas — is **not sufficient on its own**. Measured on the repro above it
takes the output from 16 tokens to 9, but 5 post-EOS tokens remain: it ends the
loop without undoing the commit.

`min_new_tokens` is unaffected: the logits processors suppress EOS until the
minimum is reached, so the target never accepts one for the trim to find (verified
at `min_new_tokens=0` → 4 tokens and `=10` → 16).

## Verification

Branch is 61 added lines (11 source, 50 test) on top of `main` (`dfb1af6`).

```
pytest tests/generation/test_utils.py tests/generation/test_candidate_generator.py \
       tests/generation/test_stopping_criteria.py
# 115 passed, 21 skipped, 3 failed
```

The 3 failures (`test_default_max_length_warning`,
`test_inputs_embeds_warn_without_ids_for_token_based_logit_processors`,
`test_validate_stopping_criteria`) fail identically on clean `main` without this
change.

The new test is armed: reverting only `generation/utils.py` to `main` makes it fail
with `generation continued past an accepted EOS: [750, 297, 297, 860, 860, 860,
860, 860]`, and it passes with the fix. `ruff check` / `ruff format --check` clean
on both touched files.

---

*AI-assisted contribution, per CONTRIBUTING.md: the patch, the repro and the test
were drafted with Claude Code, and the results above are from actual local runs.
The changed lines and the behaviour have been reviewed by me before submitting.*

---

Would you like a PR for this? I have the fix and the regression test above ready on
a branch off `main` (`dfb1af6`) — happy to open it, or to leave it if you would
rather handle the stopping logic differently.

### PR DESCRIPTION
**Pull request by pjordanandrsn**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48002)

# Crop MTP drafts at the first EOS

Fixes #47912. Companion to #47931, per @zucchini-nlp's
[request](https://github.com/huggingface/transformers/issues/47912#issuecomment-5292013856) to fix this for
MTP as well as dflash by cropping. Same crop, same place: inside the candidate generator, mirroring
`PromptLookupCandidateGenerator` (`candidate_generator.py:1129`).

## The bug

`_assisted_decoding` commits `n_matches + 1` tokens per round. It appends and streams the whole block
*before* running the stopping criteria, and `EosTokenCriteria` only inspects the last `new_token_length=1`
tokens. So an EOS accepted anywhere but the end of a committed block is missed, and generation runs past it.

## Scope: it needs more than one mtp layer

The defect fires only when the accepted EOS is *not* the final drafted token, so `num_mtp_layers == 1`
cannot reach it—the single draft position is also the last, and the `is_done_candidate` guard at
`utils.py:3777` already covers that. Every released one-layer checkpoint is therefore unaffected:
DeepSeek-V3, V3.1, GLM-4.5, 4.5-Air and 4.6 all ship exactly one. `thinkingmachines/Inkling` and
`Inkling-Small` ship eight and are reachable today.

I have no MTP checkpoint I can run (Inkling-Small is 532 GB), so the numbers below are on a tiny randomly
initialised `DeepseekV3ForCausalLM`, not on real weights.

## Why cropping the draft is sufficient

`n_matches` is the length of the leading run where the target model's argmax equals the drafted token
(`utils.py:3774`), and `valid_tokens = selected_tokens[:, : n_matches + 1]` (`:3779`). So every committed
position before the last is provably a drafted token, and only the final one—the correction/bonus token—can
be an EOS that was not in the draft. That one is already fine: it lands last, so the criteria fire. Cropping
the draft at its first EOS makes an EOS-before-the-end impossible.

The gating is exact: `EosTokenCriteria` is installed iff `generation_config._eos_token_tensor is not None`
(`utils.py:1383-1384`), and this crop fires on that same tensor, so a block can never be shortened for a
token that would not have stopped generation.

Both commit paths are covered, because both drop the bonus token once EOS lands last: greedy at
`utils.py:3777` and sampling at the matching `is_done_candidate and n_matches == candidate_length` guard in
`_speculative_sampling`. `candidate_logits` is cropped alongside `candidate_ids`—`_speculative_sampling`
indexes it positionally.

Reading `candidate_ids[0]` is safe rather than merely conventional: `_assisted_decoding` raises
`ValueError("assisted generate is only supported for batch_size = 1")` at `utils.py:3669`. The crop also
runs *before* the block is concatenated onto `input_ids`, so it only ever scans drafted tokens—an EOS in the
prompt cannot truncate a draft.

The mtp cache is untouched: the crop applies to the returned block rather than exiting the drafting loop
early, so every layer has still run and recorded its position, and the next call's
`mtp_cache.crop(-num_mtp_layers)` is driven by the layer count, never by how many candidates came back.

## Why not the `SinglePosition` shape

@zucchini-nlp suggested copying `SinglePositionMultiTokenCandidateGenerator`. That one pads post-EOS drafted
tokens with `pad_token_id` and zeroes their logits, but under the `batch_size = 1` guard its
`if sequence_stopped.all(): break` fires on the first drafted EOS, so in practice it simply stops drafting
there and the pad/zero branch is unreachable.

Ending MTP's drafting loop early would not be safe. The cache-correction path crops a fixed
`self.mtp_cache.crop(-self.num_mtp_layers)`, so a block that drafted fewer positions than there are layers
would over-crop and desync. Cropping the returned block instead leaves every layer having run and recorded
its position. Checked directly: drafting the same block with the crop firing and with
`_eos_token_tensor = None` leaves identical per-layer mtp cache lengths.

## Measured

Tiny random `DeepseekV3ForCausalLM`, `max_new_tokens=40`, on macOS/CPU/torch 2.11 and on Linux/CUDA (RTX
A2000)/torch 2.8 with identical results. A random mtp head never agrees with the main model—not one draft
accepted in 39 rounds—so these pin the shared vocab projection to make drafts accepted, which is the
precondition for the bug. Tokens committed after the first EOS, before this change:

| `num_mtp_layers` | EOS at draft index | committed after the first EOS |
|---|---|---|
| 1 | 0 (also last) | 0 |
| 2 | 0 / 1 (last) | **38** / 0 |
| 3 | 0 / 1 / 2 (last) | **38** / **37** / 0 |

Every row is 0 after it. For `num_mtp_layers=3` with the EOS at draft index 1: plain greedy emits 3 tokens
and stops at EOS, `use_mtp=True` emitted 40—37 of them after the EOS—and now emits the same 3 as greedy.
`do_sample=True` is 37 past on each of 3 seeds before and 0 after. With a draft that is cropped and then
*rejected*, generation runs 28 further rounds behind a shortened block and stays token-for-token identical
to plain greedy, which is what says the cache stays in sync.

Separately, an end-to-end A/B sweep through `generate()` (a crafted deterministic drafter, 80 CPU / 60 CUDA
cases) shows the same picture: with the crop, past-EOS is 0 and the output equals plain greedy token for
token on every case; without it, mid-block cases run 13–56 tokens past the EOS, and every
`num_mtp_layers == 1` case stays at 0 even unpatched.

## Tests

Two, in a new `TestMTPCandidateGenerator`: a tiny `DeepseekV3ForCausalLM` with five mtp layers and a pinned
`lm_head`, so the drafted tokens are exactly what each case asks for and the drafter's own weights don't
matter.

- `test_draft_is_cropped_at_the_first_eos`—EOS mid-draft, EOS *repeated* in the draft (only a repeat
  distinguishes "first" from "last"), and EOS as the very first drafted token.
- `test_draft_without_an_eos_before_the_end_is_untouched`—`_eos_token_tensor` unset, an EOS that is never
  drafted, and an EOS that is already the last drafted token.

Both assert `candidate_logits.argmax(-1)` still equals the drafted tokens, so logits cropped from the wrong
end fail; matching lengths alone would not catch that.

---

*AI-assisted contribution, per CONTRIBUTING.md: the patch, the tests and the measurements were drafted with
Claude Code, and the numbers above are from actual local runs. The changed lines and the behaviour have been
reviewed by me before submitting.*

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
can you move this inside the other PR, and lets merge it togetehr as fix EOS for candidate generators?

**Comment by pjordanandrsn** [comment]
Done—the MTP commit is now the second commit on #47931, retitled "Fix EOS for candidate generators" with the body covering both. `Fixes #47912` moved there; this one stays closed.
